cliquet/permission/__init__.py

```diff
@@ -0,0 +1,130 @@
+from cliquet.storage.exceptions import BackendError
+
+__HEARTBEAT_KEY__ = '__heartbeat__'
+
+
+class PermissionBase(object):
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def initialize_schema(self):
+        """Create every necessary objects (like tables or indices) in the
+        backend.
+
+        This is excuted with the ``cliquet migrate`` command.
+        """
+        raise NotImplementedError
+
+    def flush(self):
+        """Delete all data stored in the permission backend."""
+        raise NotImplementedError
+
+    def add_user_principal(self, user_id, principal):
+        """Add an additional principal to a user.
+
+        :param str user_id: The user_id to add the principal to.
+        :param str principal: The principal to add to the user.
+        """
+        raise NotImplementedError
+
+    def remove_user_principal(self, user_id, principal):
+        """Remove an additional principal from a user.
+
+        :param str user_id: The user_id to remove the principal to.
+        :param str principal: The principal to remove to the user.
+        """
+        raise NotImplementedError
+
+    def user_principals(self, user_id):
+        """Return the list of additionnal principals given to a user.
+
+        :param str user_id: The user_id to get the list of groups for.
+        :returns: The list of group principals the user is in.
+        :rtype: set
+
+        """
+        raise NotImplementedError
+
+    def add_object_permission_principal(self, object_id, permission,
+                                        principal):
+        """Add a principal to an Access Control Entry.
+
+        :param str object_id: The object to add the permission principal to.
+        :param str permission: The permission object to add the principal to.
+        :param str principal: The principal to add to the ACE.
+        """
+        raise NotImplementedError
+
+    def remove_object_permission_principal(self, object_id, permission,
+                                           principal):
+        """Remove a principal to an Access Control Entry.
+
+        :param str object_id: The object to remove the permission principal to.
+        :param str permission: The permission object to remove the principal to.
+        :param str principal: The principal to remove to the ACE.
+        """
+        raise NotImplementedError
+
+    def object_permission_principals(self, object_id, permission):
```

Comment:
I don't find this clearer, at least with the current name we know that it is ``object_permission_principals``, with ``get_required_principals`` we don't know what kind of principal we are talking about.

Also required let me think there are inherited but it is not the case here.